File: python/iguana/patch.py
# Copyright (c) 2026 Mario Caballero
# SPDX-License-Identifier: MIT

"""Spline patches, the maps from a parameter box into physical space"""

from __future__ import annotations

from collections.abc import Sequence

import numpy as np
import numpy.typing as npt

from iguana import cpp as _cpp


class CurvePatch:
    """A spline map from an interval into space."""

    _cpp_object: _cpp.CurvePatch

    def __init__(self, curve: _cpp.CurvePatch) -> None:
        """Initialize the curve from a C++ curve.

        Args:
            curve: A C++ curve patch.
        """
        self._cpp_object = curve

    @property
    def degree(self) -> int:
        """Polynomial degree of the curve."""
        return self._cpp_object.basis.axis(0).degree

    @property
    def knots(self) -> npt.NDArray[np.float64]:
        """Knot vector of the curve."""
        return np.asarray(self._cpp_object.basis.axis(0).knots, float)

    @property
    def control_points(self) -> npt.NDArray[np.float64]:
        """Control points, of shape ``(num_control_points, 3)``."""
        return self._cpp_object.coefficients

    def __repr__(self) -> str:
        return (f'CurvePatch(degree={self.degree}, '
                f'num_control_points={len(self.control_points)})')


class SurfacePatch:
    """A spline map from a two-dimensional parameter box into space."""

    _cpp_object: _cpp.SurfacePatch

    def __init__(self, patch: _cpp.SurfacePatch) -> None:
        """Initialize the patch from a C++ patch.

        Args:
            patch: A C++ patch object.
        """
        self._cpp_object = patch

    @property
    def degrees(self) -> tuple[int, ...]:
        """Polynomial degree of each parametric direction."""
        basis = self._cpp_object.basis

        return tuple(basis.axis(direction).degree
                     for direction in range(2))

    @property
    def knots(self) -> tuple[npt.NDArray[np.float64], ...]:
        """Knot vector of each parametric direction."""
        basis = self._cpp_object.basis

        return tuple(np.asarray(basis.axis(direction).knots, float)
                     for direction in range(2))

    @property
    def control_points(self) -> npt.NDArray[np.float64]:
        """Control points, of shape ``(num_control_points, 3)``.

        They are numbered with the first direction running fastest.
        """
        return self._cpp_object.coefficients

    def isocurves(self) -> list[CurvePatch]:
        """Isocurves of the patch along all of its knot lines.

        The curves pinned in the first direction come first, then those
        pinned in the second, each group in increasing order of knot line.
        """
        return [CurvePatch(curve) for curve in self._cpp_object.isocurves()]

    def __repr__(self) -> str:
        return (f'SurfacePatch(degrees={self.degrees}, '
                f'num_control_points={len(self.control_points)})')


class VolumePatch:
    """A spline map from a three-dimensional parameter box into space."""

    _cpp_object: _cpp.VolumePatch

    def __init__(self, patch: _cpp.VolumePatch) -> None:
        """Initialize the patch from a C++ patch.

        Args:
            patch: A C++ patch object.
        """
        self._cpp_object = patch

    @property
    def degrees(self) -> tuple[int, ...]:
        """Polynomial degree of each parametric direction."""
        basis = self._cpp_object.basis

        return tuple(basis.axis(direction).degree
                     for direction in range(3))

    @property
    def knots(self) -> tuple[npt.NDArray[np.float64], ...]:
        """Knot vector of each parametric direction."""
        basis = self._cpp_object.basis

        return tuple(np.asarray(basis.axis(direction).knots, float)
                     for direction in range(3))

    @property
    def control_points(self) -> npt.NDArray[np.float64]:
        """Control points, of shape ``(num_control_points, 3)``.

        A read-only view of the patch, not a copy. It stays valid for as
        long as the patch does.
        """
        return self._cpp_object.coefficients

    def isosurfaces(self) -> list[SurfacePatch]:
        """Isosurfaces of the patch along all of its knot planes.

        The surfaces pinned in the first direction come first, then those
        pinned in the second and third, each group in increasing order of
        knot plane. Each surface keeps the remaining two directions, in
        increasing order.
        """
        return [SurfacePatch(surface)
                for surface in self._cpp_object.isosurfaces()]

    def __repr__(self) -> str:
        return (f'VolumePatch(degrees={self.degrees}, '
                f'num_control_points={len(self.control_points)})')


def _uniform_knots(degree: int, elements: int) -> npt.NDArray[np.float64]:
    """Clamped knot vector of a uniform basis on the unit interval."""
    interior = np.linspace(0., 1., elements + 1)[1:-1]

    return np.concatenate([np.zeros(degree + 1), interior,
                           np.ones(degree + 1)])


def _greville(degree: int,
              knots: npt.NDArray[np.float64]) -> npt.NDArray[np.float64]:
    """Greville abscissa of every function of a univariate basis."""
    return np.array([knots[i + 1:i + degree + 1].mean()
                     for i in range(len(knots) - degree - 1)])


def create_box(
    lengths: Sequence[float],
    elements: Sequence[int],
    degrees: Sequence[int] = (2, 2, 2),
    origin: Sequence[float] = (0., 0., 0.),
) -> VolumePatch:
    """Create a patch mapping the parameter box onto a block.

    The block is axis-aligned and its basis uniform. The map is affine,
    so sampling it at the Greville abscissae reproduces the block
    exactly rather than approximating it.

    Args:
        lengths: Side of the block along each direction.
        elements: Number of knot spans of each direction.
        degrees: Polynomial degree of each direction.
        origin: Corner of the block the parameter box maps its own
            origin to.

    Returns:
        The patch of the block.

    Raises:
        ValueError: If an argument does not hold three entries.
    """
    given = (lengths, elements, degrees, origin)

    if any(len(argument) != 3 for argument in given):
        raise ValueError('every argument must hold three entries, one '
                         'per direction')

    knots = [_uniform_knots(degree, span)
             for degree, span in zip(degrees, elements)]

    nodes = np.meshgrid(*(_greville(degree, knot)
                          for degree, knot in zip(degrees, knots)),
                        indexing='ij')

    box = np.column_stack([node.ravel(order='F') for node in nodes])
    points = np.asarray(origin, float) + box * np.asarray(lengths, float)

    basis = _cpp.TrivariateBSpline(
        axes=[_cpp.BSpline(degree=degree, knots=list(knot))
              for degree, knot in zip(degrees, knots)])

    return VolumePatch(_cpp.VolumePatch(basis=basis, coefficients=points))

File: cpp/bindings/patch.cpp
/*
 * Copyright (c) 2026 Mario Caballero
 * SPDX-License-Identifier: MIT
 */

#include <array>
#include <cstddef>
#include <span>
#include <utility>
#include <vector>

#include <Eigen/Core>
#include <pybind11/eigen.h>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

#include "iguana/iguana.hpp"
#include "iguana/multi_index.hpp"

namespace py = pybind11;

namespace iguana::bindings
{

namespace
{

using CurvePatch = Patch<double, 1>;
using SurfacePatch = Patch<double, 2>;
using VolumePatch = Patch<double, 3>;

/**
 * @brief Axes of a basis, leaving out the one of a given direction
 *
 * @param basis Basis the axes are read from
 * @param direction Direction whose axis is left out
 *
 * @return Remaining axes, in increasing order of direction
 */
template<std::size_t d, std::size_t... index>
std::array<BSpline<double>, d - 1> other_axes(
    const TensorBSpline<double, d>& basis, std::size_t direction,
    std::index_sequence<index...>)
{
    return {basis.axis(index < direction ? index : index + 1)...};
}

/**
 * @brief Isoparametric patch at a knot line of one direction
 *
 * The result keeps the axes of the remaining directions, so that it
 * reproduces the patch on the knot line rather than sampling it. Only its
 * control points are formed, by contracting the control net against the
 * values of the pinned direction at the knot line
 *
 * @param patch Patch the isoparametric patch is read from
 * @param direction Direction pinned at the knot line
 * @param line Knot line of that direction
 *
 * @return Patch of one parametric direction less
 *
 * @pre @p line lies in [0,num_elements] of the pinned axis, which
 *      isopatches() ensures
 */
template<std::size_t d>
Patch<double, d - 1> isopatch(const Patch<double, d>& patch,
                              std::size_t direction,
                              int line)
{
    static_assert(d > 1, "isopatch: "
                         "a patch needs a direction to keep");

    const TensorBSpline<double, d>& basis = patch.basis();
    const BSpline<double>& pinned = basis.axis(direction);

    // The last line closes an element, every other one starts it
    const bool last = line == pinned.num_elements();
    const int element = last ? line - 1 : line;
    const double parameter = last ? pinned.element_end(element)
                                  : pinned.element_start(element);
    const int first = pinned.first_active(element);

    Eigen::MatrixXd weights;
    pinned.eval_on_element(first, std::span<const double>(&parameter, 1),
                           weights);

    // Function counts of the patch, and of the directions kept
    std::array<int, d> counts{};
    std::array<int, d - 1> kept_counts{};

    for (std::size_t dir = 0, kept = 0; dir < d; ++dir) {
        counts[dir] = basis.axis(dir).num_functions();

        if (dir != direction)
            kept_counts[kept++] = counts[dir];
    }

    const int num_kept = basis.num_functions() / counts[direction];
    PointMatrix<double> coefficients =
        PointMatrix<double>::Zero(num_kept, 3);

    // Every function of the net, the first direction running fastest.
    // Those active at the knot line add to the control point that shares
    // their indices in the directions kept
    std::array<int, d> index{};
    int function = 0;

    do {
        const int offset = index[direction] - first;

        if (offset >= 0 && offset < pinned.num_active()) {
            std::array<int, d - 1> kept_index{};

            for (std::size_t dir = 0, kept = 0; dir < d; ++dir) {
                if (dir != direction)
                    kept_index[kept++] = index[dir];
            }

            coefficients.row(flatten(kept_index, kept_counts)) +=
                weights(offset, 0) * patch.coefficients().row(function);
        }

        ++function;
    } while (next_lexicographic(index, counts));

    return Patch<double, d - 1>(
        TensorBSpline<double, d - 1>(other_axes(
            basis, direction, std::make_index_sequence<d - 1>{})),
        std::move(coefficients));
}

/**
 * @brief Isoparametric patches of a patch at every knot line
 *
 * @param patch Patch the isoparametric patches are read from
 *
 * @return Patches pinned in each direction in turn, at its knot lines in
 *         increasing order
 */
template<std::size_t d>
std::vector<Patch<double, d - 1>> isopatches(const Patch<double, d>& patch)
{
    std::vector<Patch<double, d - 1>> result;

    for (std::size_t direction = 0; direction < d; ++direction) {
        // n elements are bounded by the knot lines 0 to n
        const int last = patch.basis().axis(direction).num_elements();

        for (int line = 0; line <= last; ++line)
            result.push_back(isopatch(patch, direction, line));
    }

    return result;
}

} // namespace

void patch(py::module_& module)
{
    py::class_<CurvePatch>(module, "CurvePatch")
        .def_property_readonly("basis", &CurvePatch::basis)
        .def_property_readonly("coefficients", &CurvePatch::coefficients);

    py::class_<SurfacePatch>(module, "SurfacePatch")
        .def_property_readonly("basis", &SurfacePatch::basis)
        .def_property_readonly("coefficients", &SurfacePatch::coefficients)
        .def("isocurves", &isopatches<2>);

    py::class_<VolumePatch>(module, "VolumePatch")
        .def(py::init<TensorBSpline<double, 3>, PointMatrix<double>>(),
             py::arg("basis"), py::arg("coefficients"))
        .def_property_readonly("basis", &VolumePatch::basis)
        .def_property_readonly("coefficients", &VolumePatch::coefficients)
        .def("isosurfaces", &isopatches<3>);
}

} // namespace iguana::bindings
